pep517/_in_process.py

```diff
@@ -197,12 +223,23 @@ def main():
 
     hook_input = compat.read_json(pjoin(control_dir, 'input.json'))
 
+    # Add in-tree backend directories to the front of sys.path.
+    # The variable extra_path is global so that it can be used within
+    #  _build_backend.
+    global extra_path
+    extra_path = hook_input.get('backend_path')
```

Comment:
I think it would be neater to do this as an environment variable from the frontend, like we already do `PEP517_BUILD_BACKEND`, so we avoid the main() function having to put it in a global variable. That keeps all the import-related code together in `_build_backend()`.

Reply:
I'll give that a try. The only problem I have is that it's a list, so there's a translation step between list and string - I guess I can use `os.pathsep` for that, though.

Reply:
I've made this change, but in my opinion it's a lot less clean. The problem is that we need to consider all of the following possibilities:

* backend-path might not exist
* backend-path is a list, so we need to join it into a string
* encodings - path encodings are difficult, especially on Unix...

Overall, I think my original approach is better, but let me know what you think (and if you have any suggestions for making either version cleaner). Unless you're really keen on the environment variable version, I think I'll revert that and go back to the global variable.

Another option would be to make the backend a class, and the hooks methods on that class, and stop using globals altogether. But that's a bigger restructuring, and maybe should be left for a separate PR?

Reply:
I'm still inclined to prefer the version with the environment variable, but if you're really keen to go back, I'll stop arguing about it.

The only point that really presents difficulties is the encoding question, and I don't think putting it in JSON makes that any easier. Paths on Unix are fundamentally bytes, whereas JSON strings are fundamentally unicode (or UTF-16, to be more precise). Since environment variables are also bytes, we could do this neatly for Unix using `os.environb` and bytes paths. But of course that wouldn't be straightforward for Windows.

Reply:
I'm willing to leave it in (it's honestly just easier than taking it out again). But I'll note the following point for the record, it's not just (or even mostly) the encoding that's the issue here for me. It's that JSON does the job of serialising Python data structures, so that we don't have to. With JSON, we just pass the `backend_path` list and know it'll arrive at the other side unscathed. By using an environment variable, we have to do all of that serialisation ourselves.

Reply:
50% devil's advocate: the two aren't mutually exclusive. As we control both sides, we could put a JSON list in an environment variable. :wink:

Reply:
Sigh. I wondered if you'd suggest that :-P Still leaves the encoding issue, though.

I'm OK with leaving it as it is now, assuming you are. Are there any other points you think I need to address, or is this ready?

Reply:
I'd modify `sys.path` inside `_build_backend()`, rather than in `main()`. That was the main reason I preferred the environment variable: it means all the logic for finding and loading the backend can live together in `_build_backend()`.

Reply:
Ah, I see what you were meaning now - yes, that does make the whole code path a lot simpler. (I was distracted by a general principle that I heard somewhere that you should only modify `sys.path` at the start of your program - but even if that is a good principle, it makes no difference in this case).

I'll update the code on that basis, it might not be until next weekend, though.